url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/examples/book/karate/KaratePuzzle.java

```diff
@@ -0,0 +1,130 @@
+package sjm.examples.book.karate;
+
+import sjm.examples.book.combinatorics.*;
+
+/**
+ * Each of four martial arts students has a different 
+ * specialty. From the following clues, can you determine 
+ * each students full name and her special skill?
+ * <ol>
+ * <li>Ms. Ellis (whose instructor is Mr. Caldwell), Amy, and
+ *     Ms. Fowler are all martial arts students.
+ * <li>Sparring isnt the specialty of either Carla or 
+ *     Dianne.
+ * <li>Neither the shoot-fighting expert nor the pressure 
+ *     point fighter is named Fowler.
+ * <li>Childrens techniques arent the specialty of Dianne
+ *     (whose instructor is Ms. Sherman).
+ * <li>Amy, who disdains pressure point fighting, isnt 
+ *     Ms. Goodrich.
+ * <li>Betti and Ms. Fowler are roommates.
+ * <li>Ms. Hightower avoids sparring because of its point 
+ *     scoring nature.
+ * </ol>
+ * 
+ */
+public class KaratePuzzle {
+	protected Student amy = new Student("Amy");
+	protected Student betti = new Student("Betti");
+	protected Student carla = new Student("Carla");
+	protected Student dianne = new Student("Dianne");
+	protected Student[] students = { amy, betti, carla, dianne };
+	protected String[] lastNames = { "Ellis", "Fowler", "Goodrich", "Hightower" };
+	protected String[] specialties = { "Sparring", "Shoot Fighting", "Pressure Points", "Childrens" };
+
+	/*
+	 * Set the student objects' last names and specialties
+	 * from the provided arrays.
+	 */
+	protected void assembleStudents(Object[] lasts, Object[] specs) {
+		for (int i = 0; i < students.length; i++) {
+			students[i].lastName = (String) lasts[i];
+			students[i].specialty = (String) specs[i];
+		}
+	}
+
+	/**
+	 * @return true, if the student objects meet all the clues
+	 *         in the puzzle
+	 */
+	protected boolean cluesVerify() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 51-67 into a separate method.